src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,322 @@
+  RegisterState return_value_registers;
+  auto result_return_value = return_value_registers.BackupRegisters(pid);
+  FAIL_IF(result_return_value.has_error(), "Unable to read registers after function called :\"%s\"",
+          result_return_value.error().message());
+  return return_value_registers.GetGeneralPurposeRegisters()->x86_64.rax;
+}
+
+// Execute the code at `address_code`. The code segment has to end with an `int3`.
+void ExecuteOrDie(pid_t pid, RegisterState& original_registers, uint64_t address_code) {
+  RegisterState registers_set_rip = original_registers;
+  registers_set_rip.GetGeneralPurposeRegisters()->x86_64.rip = address_code;
+  auto result_restore_registers = registers_set_rip.RestoreRegisters();
+  if (result_restore_registers.has_error()) {
+    FATAL("Unable to set registers in tracee: \"%s\"", result_restore_registers.error().message());
+  }
+  if (ptrace(PTRACE_CONT, pid, 0, 0) != 0) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  int status = 0;
+  pid_t waited = waitpid(pid, &status, 0);
+  if (waited != pid || !WIFSTOPPED(status) || WSTOPSIG(status) != SIGTRAP) {
+    FATAL("Failed to wait for sigtrap after PTRACE_CONT.");
+  }
+}
+
+// Returns the absolute virtual address of a function in a module of a process as
+// FindFunctionAddress does but accepts a fallback symbol if the primary one cannot be resolved.
+ErrorMessageOr<uint64_t> FindFunctionAddressWithFallback(pid_t pid, std::string_view function,
+                                                         std::string_view module,
+                                                         std::string_view fallback_function,
+                                                         std::string_view fallback_module) {
+  ErrorMessageOr<uint64_t> primary_address_or_error = FindFunctionAddress(pid, function, module);
+  if (primary_address_or_error.has_value()) {
+    return primary_address_or_error.value();
+  }
+  ErrorMessageOr<uint64_t> fallback_address_or_error =
+      FindFunctionAddress(pid, fallback_function, fallback_module);
+  if (fallback_address_or_error.has_value()) {
+    return fallback_address_or_error.value();
+  }
+
+  return ErrorMessage(absl::StrFormat(
+      "Failed to load symbol \"%s\" from module \"%s\" with error: \"%s\"\nAnd also "
+      "failed to load fallback symbol \"%s\" from module \"%s\" with error: \"%s\"",
+      function, module, primary_address_or_error.error().message(), fallback_function,
+      fallback_module, fallback_address_or_error.error().message()));
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void*> DlopenInTracee(pid_t pid, std::filesystem::path path,
+                                                   uint32_t flag) {
+  // Figure out address of dlopen in libc.
+  OUTCOME_TRY(address_dlopen, FindFunctionAddressWithFallback(pid, "dlopen", "libdl-",
+                                                              "__libc_dlopen_mode", "libc-"));
```

Comment:
why is there a dash in the library name?

Reply:
ok, looks like I have an answer:
```
$ ls -la /usr/lib/x86_64-linux-gnu/libc.so.6   
lrwxrwxrwx 1 root root 12 Jan 14  2018 /usr/lib/x86_64-linux-gnu/libc.so.6 -> libc-2.24.so
```

Reply:
ok, I merged the change to module name now we can use "libdl.so"/"libc.so" here (it should now work for all versions).

Reply:
I'll prepare a follow up PR for this.